master/static/migrations/20240409104254_add-custom-project-key.tx.up.sql

```diff
@@ -0,0 +1,23 @@
+CREATE OR REPLACE FUNCTION function_generate_project_key(input_string TEXT)
+RETURNS TEXT AS $$
+DECLARE
+    prefix TEXT;
+    count_suffix TEXT;
+	count_value INT;
+BEGIN
+    -- Take the first 3 characters of the input string
+    prefix := LEFT(input_string, 3);
+	
+   	execute format('SELECT COUNT(*)+1 FROM projects WHERE key ILIKE $1') 
+   		into count_value
+   		using prefix || '%';
+   	count_suffix := CAST(count_value as text);
+
+    -- Concatenate prefix and count suffix
+    RETURN lower(prefix || count_suffix);
```

Comment:
I'm not too sure about this way of generating project keys. Imagine that we have a project with key `aa12`, then we have a project with name `aa1`, running this function would run into duplicated keys.

Reply:
Yea, you have a good point, I don't think the case you brought up is an immediate issue, assuming this function is only used to backfill keys for existing projects, but it will be a problem if it gets used to generate keys for new projects when there are existing keys. 

If this is only being used to backfill:
- if we have a project key of `aa12`, that would mean that:
    - this is either the 12 instance of the prefix `aa` (assuming the name of the project is `aa`)
    - or the 2nd instance of the prefix `aa1` (assuming the name of the project is `aa1`)
- so if we have a project with the prefix `aa1`
    - the count would be 3 since we only care about the keys that start with `aa1` 
    - since there is only going to be 2 or 3 other cases where that string appears as a key
        - `aa11` and `aa12` (and possibly `aa1`)

That being said, I do see a case where:
- project name: `aa1...`
- resulting key: `aa11`
- If there are 10 other projects with the name `aa` (can happen with enough workspaces)
- then we'd end up with:
    - `aa1` -> `aa11`
    - `aa` -> `aa2`
    - `aa` -> `aa3`
    - `aa` -> `aa4`
    - ... 
    - `aa` -> `aa10`
    - `aa` -> `aa11` << error on duplicate key

Now I don't think this case is as likely to happen since it's requires the project name to only be 2 characters, there to be identical 2 character projects across 10 different workspaces; but it does demonstrate it's possible.

Reply:
If we sort the way these things are handled to avoid cases like this, so instead of handling any project with `aa1...` first, we'd instead handle all the `aa` projects first, and then that would result with the  `aa1...` project being assigned the key `aa12` instead of `aa11`.

Reply:
That doesn't put us in the clear for using this style of approach for generating keys for new projects since we can't order how those will be processed or how they had been handled previously.

Reply:
For new project, since we allow passing key when creating, it would by alright to just reject duplicated key as long as we have a clear error message

Reply:
The catch is that we don't require the key to be specified, and I'm worried that if we return an error about a duplicate key that the user didn't provide it would lead to some confusion. 

The more I'm thinking about this as well, this approach also exposes potentially sensitive information about what project names (or at least the first 3 characters) exist in the system that the user making the request might not have access to. But that might be a none issue.

Reply:
The reason why I think it could be a none issue is because having incremental project keys also runs up against this, since you'll be able to tell how many projects had historically existing (they might not all exist still if they got deleted, but at some point they had).

Reply:
Suppose we have keys `a1 a2 a3` in the system, then we delete the row associated with `a1`, then a newly created project with name `a` would have the generated key to be `a3`?

Reply:
That is also an issue with this approach if used for anything outside of backfilling. If instead of counting we parse the number at the end of the key, and get the max from the set, then we could avoid that problem, but then we get into a tough spot since the generated keys will never re-use earlier values. 

Even worse, someone could specify a key of `a99999` and if we don't have way to cycle back to free keys we'll be stuck.

Reply:
We'd be in a position where we'd have to search for the free key, we could do a binary search by comparing the max & count in each subset; but this all seems like over engineering.

Reply:
If we really want something like this, I could imagine instead of basing it off an incremental value, keep the 1-3 character prefix from the project name + 2-4 random characters (w/ a retry). 

That avoids the RBAC concern outside of key collisions + avoids never re-using freed keys. This does mean that the key generation performance is probabilistic and degrades as we start hitting the upper limits of a given prefix. 

If it wasn't a product requirement to have the generated keys be based on the names, either assigning this to be 5 random alphanumeric character (or just be equivalent to the project id by default), we could avoid this problem entirely.

Reply:
Plus using the random character suffix means we don't have to enforce and order on the update & can use the same function to generate keys outside of backfill